### ISSUE TITLE
NPM@3.1.1 doesn't handle EBADPLATFORM

### ISSUE THREAD

**Issue by tivac** [npm/npm#8899]
Discovered this trying to install `webpack`, which depends on `fsevents`. `fsevents` is OSX-only, so it shouldn't have even tried.

**npm@2.13.0** [npm-debug.log](https://gist.github.com/tivac/afba8eb551a58c983ce0)
```
C:\Users\Patrick>npm i fsevents
npm ERR! Windows_NT 6.3.9600
npm ERR! argv "C:\\Program Files\\iojs\\node.exe" "C:\\Users\\Patrick\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "i" "fsevents"
npm ERR! node v2.3.1
npm ERR! npm  v2.13.0
npm ERR! code EBADPLATFORM

npm ERR! notsup Unsupported
npm ERR! notsup Not compatible with your operating system or architecture: fsevents@0.3.6
npm ERR! notsup Valid OS:    darwin
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   win32
npm ERR! notsup Actual Arch: x64
```

**npm@3.1.1** [npm-debug.log](https://gist.github.com/tivac/f3eba99294ae8a61ed18)
```
C:\Users\pcavit>npm i fsevents
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\\Program Files\\iojs\\node.exe" "C:\\Users\\pcavit\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "i" "fsevents"
npm ERR! node v2.3.4
npm ERR! npm  v3.1.1
npm ERR! code EINVALIDTYPE

npm ERR! typeerror Error: Argument #1: Expected object but got string
npm ERR! typeerror     at resolveWithNewModule (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\lib\install\deps.js:375:3)
npm ERR! typeerror     at C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\lib\install\deps.js:163:5
npm ERR! typeerror     at C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\async-map.js:52:35
npm ERR! typeerror     at Array.forEach (native)
npm ERR! typeerror     at C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\async-map.js:52:11
npm ERR! typeerror     at Array.forEach (native)
npm ERR! typeerror     at asyncMap (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\async-map.js:51:8)
npm ERR! typeerror     at exports.loadRequestedDeps (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\lib\install\deps.js:158:3)
npm ERR! typeerror     at Array.<anonymous> (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
npm ERR! typeerror     at LOOP (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:15:14)
```

**Comment by erulabs** [comment]
Can confirm - problem is resolved for me when downgrading to npm 3.1.0

Mind trying `npm install -g npm@3.1.0`?

**Comment by tivac** [comment]
@erulabs still failing for me on npm@3.1.0, though now it's failing in a new way.

I'm not set up to build native modules so it complains about python missing (which is legit). According to npm2 it shouldn't even bother trying to install at all though, so I dunno what the expected behavior here should be exactly. I suspect that matching the npm2 behavior would be best because even if I had Python I don't think building fsevents would do me much good on Windows.

**npm@3.1.0** [npm-debug.log](https://gist.github.com/tivac/b43bbd78fadd414c6f99)

```
C:\Users\pcavit>npm i -g fsevents

> fsevents@0.3.6 install C:\Users\pcavit\AppData\Roaming\npm\node_modules\fsevents
> node-gyp rebuild


C:\Users\pcavit\AppData\Roaming\npm\node_modules\fsevents>if not defined npm_config_node_gyp (node "C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node  rebuild )
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at failNoPython (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:114:14)
gyp ERR! stack     at C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:69:11
gyp ERR! stack     at FSReqWrap.oncomplete (evalmachine.<anonymous>:82:15)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "C:\\Program Files\\iojs\\node.exe" "C:\\Users\\pcavit\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\pcavit\AppData\Roaming\npm\node_modules\fsevents
gyp ERR! node -v v2.3.4
gyp ERR! node-gyp -v v2.0.1
gyp ERR! not ok
C:\Users\pcavit\AppData\Roaming\npm
└── (empty)

npm WARN EPACKAGEJSON C:\Users\pcavit\AppData\Roaming\npm No description
npm WARN EPACKAGEJSON C:\Users\pcavit\AppData\Roaming\npm No repository field.
npm WARN EPACKAGEJSON C:\Users\pcavit\AppData\Roaming\npm No README data
npm WARN EPACKAGEJSON C:\Users\pcavit\AppData\Roaming\npm No license field.
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\\Program Files\\iojs\\node.exe" "C:\\Users\\pcavit\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "i" "-g" "fsevents"
npm ERR! node v2.3.4
npm ERR! npm  v3.1.0
npm ERR! code ELIFECYCLE

npm ERR! fsevents@0.3.6 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the fsevents@0.3.6 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the fsevents package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls fsevents
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\pcavit\npm-debug.log
npm ERR! code 1
```

**Comment by tivac** [comment]
It seems like `npm@3.1.2` fixed the specific problem I had (installing `webpack`), but it's still trying to build `fsevents` on the wrong platform.

```
C:\Users\pcavit\Desktop\packer>npm i webpack

> fsevents@0.3.6 install C:\Users\pcavit\Desktop\packer\node_modules\fsevents
> node-gyp rebuild


C:\Users\pcavit\Desktop\packer\node_modules\fsevents>if not defined npm_config_node_gyp (node "C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node  rebuild )
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at failNoPython (C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:114:14)
gyp ERR! stack     at C:\Users\pcavit\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:69:11
gyp ERR! stack     at FSReqWrap.oncomplete (evalmachine.<anonymous>:82:15)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "C:\\Program Files\\iojs\\node.exe" "C:\\Users\\pcavit\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\pcavit\Desktop\packer\node_modules\fsevents
gyp ERR! node -v v2.3.4
gyp ERR! node-gyp -v v2.0.1
gyp ERR! not ok
npm WARN install:fsevents Couldn't install optional dependency: fsevents@0.3.6 install: `node-gyp rebuild`
npm WARN install:fsevents Exit status 1
C:\Users\pcavit\Desktop\packer
└─┬ webpack@1.10.1
  ├── async@0.9.2
  ├── clone@1.0.2
  ├─┬ enhanced-resolve@0.8.6
  │ └── graceful-fs@3.0.8
  ├── esprima@1.2.5
  ├── interpret@0.5.2
  ├── memory-fs@0.2.0
  ├─┬ mkdirp@0.5.1
  │ └── minimist@0.0.8
  ├── UNMET PEER DEPENDENCY node-libs-browser@>= 0.4.0 <=0.6.0
  ├─┬ optimist@0.6.1
  │ └── wordwrap@0.0.3
  ├── supports-color@1.3.1
  ├── tapable@0.1.9
  ├─┬ uglify-js@2.4.23
  │ ├── async@0.2.10
  │ ├─┬ source-map@0.1.34
  │ │ └── amdefine@1.0.0
  │ ├── uglify-to-browserify@1.0.2
  │ └─┬ yargs@3.5.4
  │   ├── camelcase@1.1.0
  │   ├── decamelize@1.0.0
  │   ├── window-size@0.1.0
  │   └── wordwrap@0.0.2
  ├─┬ watchpack@0.2.8
  │ └─┬ chokidar@1.0.5
  │   ├─┬ anymatch@1.3.0
  │   │ └─┬ micromatch@2.1.6
  │   │   ├─┬ arr-diff@1.0.1
  │   │   │ └── array-slice@0.2.3
  │   │   ├─┬ braces@1.8.0
  │   │   │ ├─┬ expand-range@1.8.1
  │   │   │ │ └─┬ fill-range@2.2.2
  │   │   │ │   ├── is-number@1.1.2
  │   │   │ │   ├── isobject@1.0.2
  │   │   │ │   ├── randomatic@1.1.0
  │   │   │ │   └── repeat-string@1.5.2
  │   │   │ ├── preserve@0.2.0
  │   │   │ └── repeat-element@1.1.2
  │   │   ├─┬ debug@2.2.0
  │   │   │ └── ms@0.7.1
  │   │   ├── expand-brackets@0.1.1
  │   │   ├── filename-regex@2.0.0
  │   │   ├── kind-of@1.1.0
  │   │   ├─┬ object.omit@0.2.1
  │   │   │ ├─┬ for-own@0.1.3
  │   │   │ │ └── for-in@0.1.4
  │   │   │ └── isobject@0.2.0
  │   │   ├─┬ parse-glob@3.0.2
  │   │   │ ├── glob-base@0.2.0
  │   │   │ ├── is-dotfile@1.0.1
  │   │   │ └── is-extglob@1.0.0
  │   │   └─┬ regex-cache@0.4.2
  │   │     ├── is-equal-shallow@0.1.3
  │   │     └── is-primitive@2.0.0
  │   ├── arrify@1.0.0
  │   ├── async-each@0.1.6
  │   ├── glob-parent@1.2.0
  │   ├─┬ is-binary-path@1.0.1
  │   │ └── binary-extensions@1.3.1
  │   ├── is-glob@1.1.3
  │   ├── path-is-absolute@1.0.0
  │   └─┬ readdirp@1.3.0
  │     ├── graceful-fs@2.0.3
  │     ├─┬ minimatch@0.2.14
  │     │ ├── lru-cache@2.6.5
  │     │ └── sigmund@1.0.1
  │     └─┬ readable-stream@1.0.33
  │       ├── core-util-is@1.0.1
  │       ├── inherits@2.0.1
  │       ├── isarray@0.0.1
  │       └── string_decoder@0.10.31
  └─┬ webpack-core@0.6.6
    ├── source-list-map@0.1.5
    └── source-map@0.4.4

npm WARN EPEERINVALID webpack@1.10.1 requires a peer of node-libs-browser@>= 0.4.0 <=0.6.0 but none was installed.
```

**Comment by iarna** [comment]
Yup, I'm tracking this over on https://github.com/npm/npm/issues/8921 now, hope to get it fixed for 3.2 this week.

**Comment by javatlacati** [comment]
@erulabs that threw me


```
module.js:472
    throw err;
    ^

Error: Cannot find module 'internal/fs'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at evalmachine.<anonymous>:18:20
    at Object.<anonymous> (C:\Users\Administrador.000\AppData\Roaming\npm\node_modules\npm\node_modules\npm-registry-client\node_modules\graceful-fs\fs.js:11:1)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
```
I'm using node 7.8.0 😞
